node/src/main/scala/org/bitcoins/node/networking/P2PClient.scala

```diff
@@ -75,58 +78,114 @@ case class P2PClientActor(
         sender() ! handleMetaMsg(metaMsg)
+      case Terminated(actor) if actor == peerConnection =>
+        context stop self
     }
 
-  /** This context is responsible for initializing a tcp connection with a peer on the bitcoin p2p network */
-  def receive: Receive =
-    LoggingReceive {
-      case cmd: Tcp.Command =>
-        //we only accept a Tcp.Connect/Tcp.Connected
-        //message to the default receive on this actor
-        //after receiving Tcp.Connected we switch to the
-        //'awaitNetworkRequest' context. This is the main
-        //execution loop for the Client actor
-        handleCommand(cmd, peerOpt = None)
-
-      case connected: Tcp.Connected =>
-        val _ = handleEvent(connected, unalignedBytes = ByteVector.empty)
-      case msg: NetworkMessage =>
-        self.forward(msg.payload)
-      case payload: NetworkPayload =>
-        logger.error(
-          s"Cannot send a message to our peer when we are not connected! payload=${payload} peer=${peer}")
+  def receive: Receive = {
+    case P2PClient.ConnectCommand =>
+      val (peerOrProxyAddress, proxyParams) =
+        peer.socks5ProxyParams match {
+          case Some(proxyParams) =>
+            val proxyAddress = proxyParams.address
+            logger.info(s"connecting to SOCKS5 proxy $proxyAddress")
+            (proxyAddress, Some(proxyParams))
+          case None =>
+            val remoteAddress = peer.socket
+            logger.info(s"connecting to $remoteAddress")
+            (peer.socket, None)
+        }
+      manager ! Tcp.Connect(peerOrProxyAddress,
+                            timeout = Some(20.seconds),
+                            options = KeepAlive(true) :: Nil,
+                            pullMode = true)
+      context become connecting(proxyParams)
+    case metaMsg: P2PClient.MetaMsg =>
+      sender() ! handleMetaMsgDisconnected(metaMsg)
+  }
 
-      case metaMsg: P2PClient.MetaMsg =>
-        sender() ! handleMetaMsg(metaMsg)
-    }
+  def connecting(proxyParams: Option[Socks5ProxyParams]): Receive = {
+    case Tcp.CommandFailed(c: Tcp.Connect) =>
+      val peerOrProxyAddress = c.remoteAddress
+      logger.error(s"connection failed to ${peerOrProxyAddress}")
+//      context stop self
+
+    case event @ Tcp.Connected(peerOrProxyAddress, _) =>
+      val connection = sender()
+      proxyParams match {
+        case Some(proxyParams) =>
+          val proxyAddress = peerOrProxyAddress
+          val remoteAddress = peer.socket
+          logger.info(s"connected to SOCKS5 proxy ${proxyAddress}")
+          logger.info(
+            s"connecting to ${remoteAddress} via SOCKS5 ${proxyAddress}")
+          val proxy = context.actorOf(
+            Socks5Connection.props(
+              sender(),
+              Socks5ProxyParams.proxyCredentials(proxyParams),
+              Socks5Connect(remoteAddress)),
+            "Socks5Connection")
+          context watch proxy
+          context become {
```

Comment:
Can you factor this out into a helper method? What is this for? A scaladoc would be nice too